### DISCUSSION TITLE
Feld: Versionierung von Titeln (Dokumenten)

### DISCUSSION THREAD

**Discussion by soerue** [OPUS4/community#32] [Ideas]
In OPUS sollten Dateien nach dem Veröffentlichen nicht mehr ausgetauscht werden. 
Es besteht aber die Option oder auch die Notwendigkeit, einen Titel zu überarbeiten und eine neue Version hochzuladen. Das heißt ein ganz neuer Datensatz in OPUS,

Dafür wäre ein Feld mit Version sinnvoll.
Version: 1.0

Eine Verlinkung oder die Möglichkeit einen Hinweis bei der vorhergehenden Version zu geben, dass es eine neue Version gibt, wäre dafür sinnvoll.

There is a [newer version](https://opus4.kobv.de/opus4-hs-XXXX/XXXX)of the record available.
Es ist eine [neuere Version] (https://opus4.kobv.de/opus4-hs-XXXX/XXXX) des Datensatzes verfügbar.

Der Versionshinweis könnte bspw. über die DOI platziert werden.
Also Format reicht vX.X
v1.0
v.1.1
v2

Vielleicht gibt es auch Meinungen ob es das kleine v Bedarf.

**Comment by j3nsch** [comment]
Für die Umsetzung stecken in dieser Anforderung eine ganze Reihe von Themen. Die Verlinkung zwischen Dokumenten wird kommen, vielleicht noch in diesem Jahr (2024). Das aufwendige dabei ist eine sinnvolle Integration ins User Interface. Man könnte einen Button einbauen, der es ermöglicht ein neues Dokument, als neue Version eines alten Dokuments zu erzeugen. Die Verlinkung könnte dabei automatisch erzeugt werden. Es sollte aber auch die Möglichkeit geben, zwei bereits existierende Dokumente, nachträglich zu verlinken. In einer ersten Version könnte man vielleicht einfach die ID des anderen Dokuments für die Verlinkung angeben, aber das geht sicherlich noch besser und bequemer. 

Die Versionsnummer wirft ein paar Fragen auf. Ein einfaches neues Feld, wäre erst einmal technisch unabhängig von der Verlinkung. Es gäbe bestimmt eine Instanz, die das Feld für andere Sachen verwenden  würde. Ist das ein Problem? Vielleicht? Es hängt davon ab, ob das Feld in OPUS 4 für weitere technische Aspekte verwendet wird. Vielleicht lassen sich die Fragen zu dem neuen Feld leichter beantworten, wenn klar ist wie die Versionsnummer verwendet werden soll. Wenn das Feld nur Zahlen, wie "1.2" aufnehmen kann, könnte man automatisch prüfen, dass die Versionsnummern der verlinkten Dokumente aufsteigend sind. Das "v" kann bei der Anzeige hinzugefügt werden, wenn gewünscht.

Die Verlinkung (in der Datenbank) selbst würde bereits klar machen, welches von zwei Dokumenten die neuere Version ist. Ein Verweis in der Datenbank mit der Bedeutung "previousVersion" würde ausreichen, um in der Frontdoor eine Verlinkung in beide Richtungen anzuzeigen. Ein zusätzlicher Verweis auf "newestVersion" wäre sinnvoll, um diese Information abrufen zu können, ohne rekursiv den Verweisen folgen zu müssen. Dieser Verweis müsste beim Anlegen einer neuen Version für alle verlinkten Dokumente aktualisiert werden. Das wäre ein einfaches SQL-Update. Generell muss dafür gesorgt werden, dass die Verlinkung aktualisiert wird, wenn z.B. ein Dokument in der Kette gelöscht wird. Das sollte im Normalfall nicht passieren, aber das System muss damit umgehen können. 

Das für die Verlinkung notwendige System, könnte auch für andere Beziehung zwischen Dokumenten in einem OPUS 4 System verwendet werden.

In OPUS 4.8.1 ist das Feld `PublicationState` aktiviert und wird mit dem Label "Version" angezeigt. Was wäre ein guter Name für das neue Version-Feld, intern und im User Interface? Vielleicht sollten die Links und die Version zusammen mit `PublicationState` in einem separaten Bereich im Metadaten-Formular angezeigt werden? Für `PublicationState` gibt es ja unter anderem die Werte "enhancedVersion" und "updatedVersion", die dann im Zusammenhang mit einer Versionierung in OPUS 4 noch mehr Sinn ergeben. 

Die Umsetzung kann sicherlich in mehreren Schritten erfolgen. Was ist der minimale erste Schritt? Danach gibt es aber mit Sicherheit weitere Auswirkungen. Wie sollen mehrere Versionen bei der Suche behandelt werden. Gibt es Folgen für OAI-PMH, für die verschiedenen Export-Formate? Woran muss noch gedacht werden?

**Comment by soerue** [comment]
Ich habe dies als "Idee" gekennzeichnet, **da es keine hohe Priorität darstellt.**

Bei PublicationState würde ich nicht die Versionierungsangabe verstehen, sondern:
ob es live ist oder preview / draft oder published.
Bzw. PublicationStatus

Zu der Versionierung - hier würde meiner Meinung nach, erstmal das Feld reichen.

Da Verlinkungsmöglichkeiten geben soll - bzw. aktive Links - wäre die Umsetzung erstmal anzuschauen, welche Links das jetzt sind.
Über das Bemerkungsfeld haben wir bei unser Instanz Informationen bspw. zu zugehörenden Material verlinkt. Wäre da die Verlinkung aktiv, wäre ein Versionierungshinweis auch dort möglich.
Normalerweise, werden die Versionsangaben, meist rechts angezeigt. Oder eine Information als Banner.

Ich habe mal in der DataCite Dokumentation nachgesehen.

DataCite documentation

This field is optional.
Note that this field refers to the version of the resource, not the version of the metadata.
The versioning information is set according to the policies of the data provider (data publishers do not change/use this field).
It is recommended to use semantic versioning [external link]. Up to three labels are supported (Major, Minor, Patch). Depending on the resource, only one or two labels might be needed.
Example 1 (two labels - Major.Minor)
\<version>4.2\</version>
Example 2 (three labels - Major.Minor.Patch)
\<version>4.2.1\</version>


Oder im OAI-Record:
\<resource xsi:schemaLocation="http://datacite.org/schema/kernel-4 http://schema.datacite.org/meta/kernel-4.3/metadata.xsd" >
\<creators>
\<creator>
\<creatorName nameType="Personal" >Li, Yunxiao\</creatorName>
\<givenName>Yunxiao\</givenName>
\<familyName>Li\\</familyName>
\<nameIdentifier nameIdentifierScheme="ORCID" >0000-0002-5537-3774\</nameIdentifier>
\<affiliation>University of California Riverside\</affiliation>
\</creator>
\</creators>
\<titles>
\<title>Data accompanying the GitHub repository bartonlab/paper-clonal-dynamics\</title>
\</titles>
\<publisher>XXX\</publisher>
\<publicationYear>2023\</publicationYear>
\<dates>
\<date dateType="Issued" >2023-10-12\</date>
\<date dateType="Updated" >2023-10-12\</date>
\</dates>
\<resourceType resourceTypeGeneral="Dataset" >\</resourceType>
\<relatedIdentifiers>
\<relatedIdentifier relatedIdentifierType="DOI" relationType="IsVersionOf" >10.5281/XXX\</relatedIdentifier>
\</relatedIdentifiers>
**\<version>1.0.0\</version>**
\<rightsList>
\<rights rightsURI="https://creativecommons.org/licenses/by/4.0/legalcode" rightsIdentifierScheme="spdx" rightsIdentifier="cc-by-4.0" >Creative Commons Attribution 4.0 International\</rights>
\</rightsList>
\<descriptions>
\<description descriptionType="Abstract" >This dataset contains data that accompany the GitHub repository bartonlab/paper-covariance-estimation. The GitHub repository contains code and scripts for reproducing results described in the manuscript "Correlated allele frequency changes reveal clonal structure and selection in temporal genetic data." See the GitHub repository for details on the interpretation and analysis of this data.</description>
\</descriptions>
\</resource>

**Reply by j3nsch** [reply]
Vielen Dank! 

> Da Verlinkungsmöglichkeiten geben soll - bzw. aktive Links - wäre die Umsetzung erstmal anzuschauen, welche Links das jetzt sind.
> Über das Bemerkungsfeld haben wir bei unser Instanz Informationen bspw. zu zugehörenden Material verlinkt. Wäre da die Verlinkung aktiv, wäre ein Versionierungshinweis auch dort möglich.
> Normalerweise, werden die Versionsangaben, meist rechts angezeigt. Oder eine Information als Banner.

Bei der "Verlinkung" geht es mir hier konkret, um Beziehungen zwischen Dokumenten in einer OPUS 4 Instanz. Welche Verknüpfungen möglich sein werden, hängt von den Anforderungen ab. Es geht jetzt also darum zu definieren was mindestens möglich sein muss. 

Die Ausgabe von URLs in Bemerkungen und anderen Feldern als anklickbare Links, ist davon unabhängig. Prinzipiell ist das jetzt schon möglich. Im Standard ist es momentan nicht vorgesehen und es gibt keinen einfachen Schalter, den man umlegen kann, um es zu aktivieren. 
Insbesondere wenn Bemerkungen von unbekannten Nutzer eingegeben werden können, muss man damit rechnen, das unerwünschtes HTML/Javascript ausgegeben werden könnte. 

Es geht mir hier auch nicht um Links zu externen Ressourcen. Dafür wären **References** sinnvoll, die momentan in OPUS 4 nicht aktiv sind.  Die **Identifier** in OPUS 4 sind als alternative IDs für ein Dokument gedacht. Die **References** sind als IDs, z.B. URLs, für andere Ressourcen gedacht, die mit dem Dokument zu tun haben. In der Praxis, werden Identifier teilweise für andere Ressourcen verwendet. Das Datenmodell für Identifier und References wird sich ändern. Das ist aber ein größeres separates Thema.

Zur **Priorisierung** kann ich nur sagen, dass bei vielen Dingen die technische Umsetzung nicht das Hauptproblem ist, sondern die Fragen, die sich aus der Umsetzung ergeben. Häufig scheint der einzige Weg, detaillierte  Anforderungen zu bekommen, etwas zu programmieren, dass nicht funktioniert und damit Feedback zu provozieren, wie es funktionieren sollte. Je mehr vorab geklärt werden kann, desto besser.
Die Möglichkeit Dokumente in Beziehungen zueinander zu setzen, ist für andere Features wichtig, zum Beispiel, um Dubletten markieren zu können.

Wenn ein Feld erst einmal reichen würde, könnte dafür auch jetzt schon ein Enrichment verwendet werden. Die reine Erfassung reicht aber ja nicht und es sollten hier möglichst alle kleinen und großen Teilaufgaben gesammelt werden, die noch notwendig sind.

**Comment by j3nsch** [comment]
@alw-bsz Ideen wir man den Konflikt mit "Version" (PublicationState) lösen könnte, damit es keine Verwirrungen gibt? 

Das Feld würde ich intern einfach **Version** bzw. evtl. **VersionNumber** nennen. Das `DocumentInterface` müsste um Funktionen zum Hinzufügen und Entfernen von Versionen erweitert werden, die sich im Hintergrund um die Verlinkung kümmern. Die Versionsnummer ist im Prinzip davon unabhängig und würde nur für die Anzeige bzw. Ausgabe in Export-Formaten verwendet werden. Die "Previous/Next/Current/All"-Versions (Dokumente) sind einzig durch die Verlinkung ermittelbar.

**Comment by alw-bsz** [comment]
Ich habe folgende Ideen für die deutsche Bezeichnung:

Entweder man belässt für _PublicationState_ die Übersetzung "Version"; dann würde ich für das neue Feld _VersionNumber_ zu "Versionszählung" tendieren. Oder man benennt _PublicationsState_ um (am naheliegendsten wäre wohl "Fassung"); dann könnte man für _VersionNumber_ wahlweise "Version" oder "Versionszählung" verwenden.

Im Englischen würde ich bei "Version" für _PublicationState_ bleiben und _VersionNumber_ mit "Version Number" bezeichnen.
